src/pool/srv_target.c

```diff
@@ -1094,113 +1090,171 @@ ds_pool_put(struct ds_pool *pool)
+{
+	struct cont_srv_open_arg *arg = vin;
+
+	return ds_cont_srv_open(arg->pool_uuid, arg->cont_hdl_uuid, arg->flags, arg->sec_capas);
+}
+
+int
+ds_pool_srv_open(uuid_t pool_uuid, uuid_t cont_hdl_uuid)
+{
+	struct cont_srv_open_arg arg = {0};
+
+	D_ASSERT(!uuid_is_null(cont_hdl_uuid));
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	uuid_copy(arg.cont_hdl_uuid, cont_hdl_uuid);
+	arg.flags     = 0;
+	arg.sec_capas = ds_sec_get_rebuild_cont_capabilities();
+
+	return ds_pool_task_collective(pool_uuid,
+				       PO_COMP_ST_NEW | PO_COMP_ST_DOWN | PO_COMP_ST_DOWNOUT,
+				       cont_srv_open_one, &arg, 0);
+}
+
+static inline bool
+eph_report_exiting(struct ds_pool *pool)
+{
+	return pool->sp_stopping || dss_ult_exiting(pool->sp_ec_ephs_req);
+}
+
+#define EPH_REPORT_INTVL       5000 /* 5 seconds */
+#define EPH_REPORT_RETRY_INTVL 2000 /* 2 seconds */
+
+static void
+eph_report_ult(void *data)
 {
 	struct ds_pool	*pool = data;
-	int		rc = 0;
+	int              rc, sleep_intvl;
+	bool             conn_hdl_fetched = false, srv_hdl_fetched = false;
 
-	D_INFO(DF_UUID": begin: fetch_hdls=%u stopping=%u\n", DP_UUID(pool->sp_uuid),
-	       pool->sp_fetch_hdls, pool->sp_stopping);
+	D_DEBUG(DB_MD, DF_UUID " Enter eph report.\n", DP_UUID(pool->sp_uuid));
+	D_ASSERT(pool->sp_ec_ephs_req != NULL);
+
+	if (pool->sp_stopping)
+		goto out;
 
 	/* sp_map == NULL means the IV ns is not setup yet, i.e.
 	 * the pool leader does not broadcast the pool map to the
 	 * current node yet, see pool_iv_pre_sync().
 	 */
 	ABT_mutex_lock(pool->sp_mutex);
 	if (pool->sp_map == NULL) {
-		D_INFO(DF_UUID": waiting for map\n", DP_UUID(pool->sp_uuid));
+		D_INFO(DF_UUID ": Wait for pool map.\n", DP_UUID(pool->sp_uuid));
 		ABT_cond_wait(pool->sp_fetch_hdls_cond, pool->sp_mutex);
+		D_INFO(DF_UUID ": Wait for pool map done.\n", DP_UUID(pool->sp_uuid));
 	}
 	ABT_mutex_unlock(pool->sp_mutex);
 
-	if (pool->sp_stopping) {
-		D_DEBUG(DB_MD, DF_UUID": skip fetching hdl due to stop\n",
-			DP_UUID(pool->sp_uuid));
-		D_GOTO(out, rc);
-	}
-	D_INFO(DF_UUID": fetching handles\n", DP_UUID(pool->sp_uuid));
-	rc = ds_pool_iv_conn_hdl_fetch(pool);
-	if (rc) {
-		D_INFO(DF_UUID" iv conn fetch %d\n", DP_UUID(pool->sp_uuid), rc);
-		D_GOTO(out, rc);
-	}
+	while (!eph_report_exiting(pool)) {
+		sleep_intvl = EPH_REPORT_INTVL;
+
+		/* Fetch pool connection handles */
+		if (!conn_hdl_fetched) {
+			D_INFO(DF_UUID ": Fetching connection handles.\n", DP_UUID(pool->sp_uuid));
+			rc = ds_pool_iv_conn_hdl_fetch(pool);
+			if (rc) {
+				D_INFO(DF_UUID ": Failed to fetch connection handles. " DF_RC "",
+				       DP_UUID(pool->sp_uuid), DP_RC(rc));
+				sleep_intvl = EPH_REPORT_RETRY_INTVL;
+			} else {
+				conn_hdl_fetched = true;
+			}
+		}
 
-out:
-	D_INFO(DF_UUID": signaling done\n", DP_UUID(pool->sp_uuid));
-	ABT_mutex_lock(pool->sp_mutex);
-	ABT_cond_signal(pool->sp_fetch_hdls_done_cond);
-	ABT_mutex_unlock(pool->sp_mutex);
+		if (eph_report_exiting(pool))
+			break;
```

Comment:
Move the check `eph_report_exiting` inside `if (!conn_hdl_fetched)` section to avoid unnecessary check for each `while` cycle.

Reply:
sure